azkaban-common/src/main/java/azkaban/scheduler/MissedSchedulesManager.java

```diff
@@ -0,0 +1,255 @@
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package azkaban.scheduler;
+
+import azkaban.Constants;
+import azkaban.project.Project;
+import azkaban.project.ProjectManager;
+import azkaban.trigger.builtin.ExecuteFlowAction;
+import azkaban.utils.DaemonThreadFactory;
+import azkaban.utils.Emailer;
+import azkaban.utils.Props;
+import com.google.common.annotations.VisibleForTesting;
+import com.google.inject.Inject;
+import com.google.inject.Singleton;
+import java.text.MessageFormat;
+import java.util.ArrayList;
+import java.util.Date;
+import java.util.List;
+import java.util.Queue;
+import java.util.concurrent.ConcurrentLinkedQueue;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.Future;
+import java.util.concurrent.TimeUnit;
+import java.util.stream.Collectors;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Miss Schedule Manager is a failure recover manager for schedules. Schedule might be missed to execute due to
+ * Trigger Scanner thread busy or web server down. This manager maintains a configurable fixed sized thread pool
+ * to execute tasks including send email notification or back execute flow.
+ * */
+@Singleton
+public class MissedSchedulesManager {
+  private static final Logger LOG = LoggerFactory.getLogger(MissedSchedulesManager.class);
+
+  private final ExecutorService missedSchedulesExecutor;
+  private final List<MissedSchedulesOperationTask> backgroundTasksList;
+  private final Queue<MissedScheduleTaskQueueNode> missedScheduleTaskQueue = new ConcurrentLinkedQueue<>();
+  private final ProjectManager projectManager;
+  private final Emailer emailer;
+
+  // parameters to adjust MissScheduler
+  private final long sleepIntervalInMs;
+  private final long DEFAULT_THREAD_IDLE_TIME_IN_MS = 60000;
+  private final int threadPoolSize;
+  private final int DEFAULT_THREAD_POOL_SIZE = 50;
+
+  @Inject
+  public MissedSchedulesManager(final Props azkProps,
+      final ProjectManager projectManager,
+      final Emailer emailer) {
+    this.projectManager = projectManager;
+    this.emailer = emailer;
+    this.sleepIntervalInMs = azkProps.getLong(Constants.MISSED_SCHEDULE_THREAD_IDLE_TIME, DEFAULT_THREAD_IDLE_TIME_IN_MS);
+    this.threadPoolSize = azkProps.getInt(Constants.MISSED_SCHEDULE_THREAD_POOL_SIZE, DEFAULT_THREAD_POOL_SIZE);
+    if (!azkProps.getBoolean(Constants.MISSED_SCHEDULE_MANAGER_ENABLED, false) || threadPoolSize <= 0) {
```

Comment:
Why having a feature flag property if the same result can be achieved by setting thread pool size to a negative value? Why not fail initialization of MissedSchedulesManager if the thread pool size configured is negative? What's the advantage of allowing negative values?